### ISSUE TITLE
[dependency-injection] Thoughts on the PhpDumper 

### ISSUE THREAD

**Issue by eFrane** [symfony/symfony#37984]
**Description**  

Disclaimer: I have no clue whether this should be an issue or if this is the right issue type. Sorry in advance for any mistakes. Also, please help me with the title 🙈.

tl;dr: I propose to change the way the PhpDumper - and partially container dumping in general - works.

**Ideas**

_Doing this as a list, if doing any if this is to be considered at all, separate issues should probably be opened_

1) Use a code generator (e.g. `nette/php-generator`) instead of `sprintf` for the file generation
2) Make lookup paths (pretty much everything `PhpDumper::export()` does behind the scenes) less magical
3) Allow setting the dumper or passing dumper options during container dump (`MicroKernelTrait::dumpContainer()`)

**Example**  

_No examples for two reasons: All of these suggestions would not surface to the regular end user and I don't want to bloat this too long issue description with numerous examples unless there's interest._

**Comment by wouterj** [comment]
Thank you for creating an issue with your thoughts :). I think it would help the core team if you explain *why* you think these changes would be needed. This way, it can be compared to the costs (of making the changes & resolving merge conflicts during the release lifetime, etc). As it's currently written, I fail to see much advantages other than making a very internal class less magical.

**Comment by eFrane** [comment]
Yeah, I can see how amending the _why_ might be useful. 

For starters, I do think that making a very internal class less magical would be a good thing in and of itself. Reasons for doing so aside, it's not very welcoming of any codebase to "require" hours of head scratching when one accidentally stumbles into internal classes on the quest to understand why certain code blocks do what they do.

_Why tho?_. Well, this might lead to this issue deviating into a support question which wasn't my intention, but here goes. While trying to convince a skeleton application to run inside a Phar, I stumbled upon the problem of (obviously) not being able to compile the container inside the read-only filesystem of said Phar. The route I am currently going is to build the container "by hand", that is I set up a separate ConfigCache to be included in the Phar, initialize and compile my Kernel and load the cached container when starting inside the Phar. That works well enough. However, I keep having the weirdest path lookup issues which eventually led me into the `PhpDumper` and subsequently into not at all understanding why there are enforcements on the number of root directories or why there is no option to say "no, do not resolve to the current vendor directory". (This last one is funny because in my case it means the generated Phar will very well work inside the directory it is created in but will utterly fail finding files when started in any other location.)

Anyway, amending to my initial listing:

> 1) Use a code generator (e.g. nette/php-generator) instead of sprintf for the file generation

I have learned so much from the mostly truly wonderful code in Symfony, especially in the more internal classes. Seeing so much `sprintf` and Heredoc to generate PHP honestly just feels out of place. Correct me if I'm wrong but my impression always was that Symfony is as much about Developer Experience for it's users as it is about well-maintained and high-quality code on the inside. 

I get that adding an external dependency for code generation would be a big decision, not only would it mean pretty much rewriting the whole dumper but well, external dependencies add costs. However, I'd say this could also be improved upon without adding a code generator like I initially suggested.

> 2) Make lookup paths (pretty much everything `PhpDumper::export()` does behind the scenes) less magical

tl;dr: See my Phar adventure above. It's highly likely I'm barking up the wrong tree here but the general point still stands. In any case, this is mostly me being unsatisfied with the readability of this method. It's highly likely that some sprinkled code commentary would do wonders for the occasional wanderer stumbling upon this. So, yeah, readability is a big one for me here again.

> 3) Allow setting the dumper or passing dumper options during container dump (`MicroKernelTrait::dumpContainer()`)

First, a correction which you probably already spotted, `dumpContainer` is of course not in the `MicroKernelTrait`. Not that that is important for the issue. This would be the big one. I understand that DI container setup and everything involved with it is probably one of the most valuable parts of what makes Symfony Symfony which makes changing these code sections hard to justify. Especially as touching anything nearby `Kernel::configureContainer()` or `::build()` would highly likely be an upgrade issue for numerous applications. 

However – and again, I might be utterly wrong here – I think passing the Dumper into `Kernel::dumpContainer()`, or at least allowing custom configuration with e.g. a `protected Kernel::getContainerDumper(): DumperInterface` method would in my opinion offer a customisation point for the crazy ones like me without really changing anything for the remaining 99%. 

I hope I could make my intentions a little clearer.

**Comment by nicolas-grekas** [comment]
Thanks for the details.

> Use a code generator (e.g. nette/php-generator) instead of sprintf for the file generation

I've been working on the class a lot, and from my experience its complexity doesn't come from the sprintf() at all. Yes, there is indentation logic that would be outsourced to a codegen, but overall, the pros/cons balance of adding a dependency (both in terms of adoption cost + expected benefit) are not worth it.

> Make lookup paths (pretty much everything PhpDumper::export() does behind the scenes) less magical

That's a good call. But we don't have any concept of "path" anywhere that would allow making this logic around path "less magical". I'm very open to proof of concepts if you're up to digging into this (this won't be trivial, if possible).

> Allow setting the dumper or passing dumper options during container dump (MicroKernelTrait::dumpContainer())

On this one, I think the code might be already flexible enough: your use case is really nice, but it's really rare also. Since dumpContainer is protected, you're free to override it at will. Don't get me wrong: adding arguments to pass options is possible and we would consider a PR doing so (this should be implemented in a BC way, but review will help achieve this anyway)

**Comment by eFrane** [comment]
Thanks for the insights @nicolas-grekas and @wouterj. I will close this issue and likely return with either more concrete ones or – if all goes well – pull requests to further address my nitpickings.
